zasady_konstrukcji_api.rst

```diff
@@ -0,0 +1,383 @@
+- ścieżka do korzenia udostępnia index API.
+
+W odróżnieniu, niektóre aplikacje o architekturze SOA (service–oriented architecture) komunikują się sztywnym interfejsem albo
+językiem opisu interfejsu (IDL).
+
+Zasada HATEOAS rozdziela klienta i serwer w sposób, który pozwala im ewoluować niezależnie.
+
+Umożliwiaj filtrowanie, sortowanie, wybór pól, stronicowanie przy pobieraniu kolekcji
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Elastyczność filtrowania rozluźnia związek API z modelem.
+
+Przykłady
+
+::
+
+  GET /cars?color=red            Lista czerwonych samochodów
+
+  GET /cars?seats<=2             Lista samochodów o max 2 siedzeniach
+
+  GET /cars?sort=-manufacturer,+model
+
+  GET /cars?fields=manufacturer,model,id,color
+
+  GET /cars?offset=10&limit=5
+
+Udzielaj kompletnych odpowiedzi
++++++++++++++++++++++++++++++++
+
+Odpowiedzi powinny zawierać wszystkie dane potrzebne do utworzenia kompletnego obiektu modelu.
+
+::
+
+  // GET /users/12
+  {
+    user: {
+      id: 12,                    // powtórzone z pytania, ale ułatwia pracę klienta
+      name: 'Bob Barker'
+    }
+  }
+
+  // GET /users/12
+  {
+    user: {
+      id: 12,
+      name: 'Bob Barker',
+      todo_list_ids: [ 4, 5 ]    // struktury będą w osobnych rekordach; dyskusyjne
+    }
+    todo_lists: [
+      { id: 4, name: 'shopping' },
+      { id: 5, name: 'work' }
+    ]
+  }
+
+Używaj warstwy prezentacji
+++++++++++++++++++++++++++
+
+Separuj dane API od warstwy danych aplikacji:
+
+- pozwala obliczać wtórne wartości, które mogą być potrzebne po stronie klienckiej, np. obliczenie płci na podstawie numeru PESEL
+- tworzy to dobrą warstwę do testów
+- chroni kod kliencki od zmian po stronie serwera.
+
+Daj przykłady odpowiedzi na GET
++++++++++++++++++++++++++++++++
+
+Pokaż, jakich danych można oczekiwać w odpowiedzi na wszystkie poprawne rodzaje żądań GET.
+Przykłady powinne być proste i łatwe do zrozumienia
+(czyli w czasie < 5 sek.).
+
+Umożliw ograniczenie listy zwracanych pól
++++++++++++++++++++++++++++++++++++++++++
+
+Parametr taki jak „pola”, podający rozdzieloną przecinkami listę, pozwala na ograniczenie zwracanych pól do tych,
+których użytkownik potrzebuje. Na przykład
+
+::
+
+  GET /tickets?fields=id,subject,customer_name,updated_at
+
+Tworzenie i aktualizacje powinny zwracać nowe dane
+++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Operacje PUT, POST i PATCH mogą dokonywać zmian pól, które nie były dostarczone na liście parametrów (np. czas).
+
+Niektórzy autorzy postulują, że
+aby uniknąć konieczności ponownego pytania o zaktualizowane wartości, operacja aktualizująca powinna od razu
+je zwracać. (To jest dyskusyjne, np. w przypadku masowych zmian — których efektu program kliencki może nie potrzebować.)
+
+Zwracaj JSON
+++++++++++++
+
+XML jest zbyt dosłowny i trudniejszy do czytania i parsowania.
+
+W sieci obserwuje się odejście od XML na rzecz formatu JSON.
```

Comment:
O tyle to zalecenie jest problematyczne, że XML jest często wykorzystywany w polskiej administracji i niekiedy będzie naturalnie zwracanym formatem. Chociażby:

* UPP z ePUAP jest zwracany w XML;
* pisma w ePUAP są zapisywane w XML;
* akty prawne są zapisywane w XML;
* paczka archiwalna korzysta z formatu XML.

Być może warto jakoś rozróżnić sytuacje, w których XML jest akceptowalnym rozwiązaniem i takie, gdzie JSON będzie wskazany.